### PR TITLE
Allow migration of VMs with host-model CPU to migrate only for compatible nodes

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
The PR addresses [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1760028).

VMs with "host-model" CPU should not be able to migrate to other nodes that do not support the CPU model. The problem is that for a running VMI the migration controller (which schedules the cloned vmi pod on another node) cannot know the actual model that is used for host-cpu model. Therefore now node-labeller adds another label to every node in the cluster which states the actual CPU model that is used for "host model".

This way the migration controller can ask to schedule the new pod only on nodes that support the CPU model. If no such nodes exist the migration will fail.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1760028

**Release note**:

```release-note
[bugfix]: Allow migration of VMs with host-model CPU to migrate only for compatible nodes
```

### CODE REVIEW

**Review by davidvossel — COMMENTED** [review]
seems reasonable. i made a comment about using node cache, but the general approach makes sense to me

/cc @vladikr Can you take a look at this as well?

File: pkg/virt-controller/watch/migration.go [hunk 1]
```diff
@@ -466,6 +467,19 @@ func (c *MigrationController) createTargetPod(migration *virtv1.VirtualMachineIn
 	// TODO libvirt requires unique host names for each target and source
 	templatePod.Spec.Hostname = ""
 
+	// If cpu model is "host model" allow migration only to nodes that supports this cpu model
+	if cpu := vmi.Spec.Domain.CPU; cpu != nil && cpu.Model == virtv1.CPUModeHostModel {
+		node, err := c.clientset.CoreV1().Nodes().Get(context.Background(), vmi.Status.NodeName, v1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:472]: Thanks for the important note - didn't know about informers, totally makes sense.

File: pkg/virt-controller/watch/migration.go [hunk 2]
```diff
@@ -106,6 +110,12 @@ func NewMigrationController(templateService services.TemplateService,
 		UpdateFunc: c.updateMigration,
 	})
 
+	c.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:113]: Do you want to do this?  You might only want the `nodeInformer` for cache, not to trigger the `sync`.

File: pkg/virt-handler/node-labeller/cpu_plugin.go [hunk 3]
```diff
@@ -26,47 +26,63 @@ import (
...
 	util "kubevirt.io/kubevirt/pkg/virt-handler/node-labeller/util"
 )
 
 const (
 	isUnusable             string = "no"
+	isRequired             string = "require"
 	nodeLabellerVolumePath        = "/var/lib/kubevirt-node-labeller/"
+
+	supportedFeaturesXml = "supported_features.xml"
+	x86PenrynXml         = "x86_Penryn.xml"
```
Comment (xpivarc) [inline pkg/virt-handler/node-labeller/cpu_plugin.go:42]: Any reason to have this here? I see it only in tests.

File: tests/migration_test.go [hunk 4]
```diff
@@ -1786,6 +1786,114 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
...
 		})
+		Context("with a host-model cpu", func() {
+			getNodeHostModel := func(node *k8sv1.Node) (hostModel string) {
+				for key, _ := range node.Labels {
+					if strings.HasPrefix(key, v1.HostModelCPULabel) {
+						hostModel = strings.TrimPrefix(key, v1.HostModelCPULabel)
+						break
+					}
+				}
+				Expect(hostModel).ToNot(BeEmpty()) // Must find node's host model
```
Comment (xpivarc) [inline tests/migration_test.go:1797]: You can move the comment to the description of the assert. Then we can see what's wrong without looking into the code.

File: tests/migration_test.go [hunk 5]
```diff
@@ -1786,6 +1786,114 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
...
+							return true
+						}
+					}
+				}
+
+				return false
+			}
+
+			It("should migrate only to nodes supporting right cpu model", func() {
+				if !toPerformTest() {
```
Comment (xpivarc) [inline tests/migration_test.go:1866]: Can we name this a little bit better? Maybe `isHeterogeneousCluster` ?

**Review by xpivarc — COMMENTED** [review]
Looks good, you probably want only cache and not to sync on node`s changes.

File: pkg/virt-controller/watch/migration.go (hunk 2 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:113]: > You might only want the `nodeInformer` for cache, not to trigger the `sync`.

I'm not sure I understand what you mean.
If I delete the [run command to sync caches](https://github.com/kubevirt/kubevirt/blob/9a5075928111de2b840d31d9d8b8ee21930b5b25/pkg/virt-controller/watch/migration_test.go#L154) the test hangs forever and fails on timeout. Also if I don't refer to events regarding nodes (node creation / deletion etc') what will notify the the work-queue to get the relevant node into the informer's cache?

Do you maybe mean that there should only be an `addFunc` defined for `AddEventHandler()` (without `DeleteFunc`/`UpdateFunc`)? I can delete the `UpdateFunc`, but wouldn't it be problematic to remove `DeleteFunc` since we then provide a node from cache that has been deleted?

I'm new to all of this so probably missing something :)

File: pkg/virt-handler/node-labeller/cpu_plugin.go (hunk 3 above)
Comment (iholder-redhat) [reply pkg/virt-handler/node-labeller/cpu_plugin.go:42]: Thanks! :smile: 
Indeed `x86PenrynXml` is used only by `cpu_plugin_test.go` and therefore moved to there :)

File: pkg/virt-controller/watch/migration.go (hunk 2 above)
Comment (xpivarc) [reply pkg/virt-controller/watch/migration.go:113]: The informer store the objects to cache. You don't need to do anything else.
[See also ](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L557)

`EventHandler` is for the case when you want to enqueue the object to the controller work-queue. Upon the work-queue the controller syncs the observed and desired state.

So I would propose to remove the registration of the handler.

File: pkg/virt-controller/watch/migration.go [hunk 6]
```diff
@@ -948,6 +968,25 @@ func (c *MigrationController) deleteVMI(obj interface{}) {
 	}
 }
 
+func (c *MigrationController) addNode(obj interface{}) {
```
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:971]: This is not used, right now.

**Review by xpivarc — APPROVED** [review]

File: pkg/virt-controller/watch/migration.go [hunk 7]
```diff
@@ -466,6 +470,22 @@ func (c *MigrationController) createTargetPod(migration *virtv1.VirtualMachineIn
...
+	if cpu := vmi.Spec.Domain.CPU; cpu != nil && cpu.Model == virtv1.CPUModeHostModel {
+		obj, exists, err := c.nodeInformer.GetStore().GetByKey(vmi.Status.NodeName)
+		node := obj.(*k8sv1.Node)
+		if err != nil {
+			return fmt.Errorf("cannot get nodes to migrate VMI with host-model CPU. error: %v", err)
+		} else if !exists {
+			return fmt.Errorf("node \"%s\" assosiated with vmi \"%s\" does not exist", vmi.Status.NodeName, vmi.Name)
+		}
+
+		err = prepareNodeSelectorForHostCpuModel(node, templatePod)
```
Comment (vladikr) [inline pkg/virt-controller/watch/migration.go:483]: I think we should fail the migration if we can't find the host model labels on the node. There is nothing really we can wait for if these labels don't exist on the node.

**Review by vladikr — COMMENTED** [review]
Great work! Great to see this is coming together. 
From my side, I think we are missing a feature gate. I think we need to block migrations with host-model and only allow migration if the feature gate is enabled.

I see that HCO[1] sets a non-existing feature-gate `WithHostModelCPU` for some reason... 
We can actually implement it.

I think that we also should fail the migration early ([somewhere here](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/watch/migration.go#L399)) in the case that no host-model labels were set on a source node. 
[1]https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/controller/operands/kubevirt.go#L108`

File: pkg/virt-controller/watch/migration.go (hunk 7 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:506]: > Great work! Great to see this is coming together.
> From my side, I think we are missing a feature gate. I think we need to block migrations with host-model and only allow migration if the feature gate is enabled.
> 
> I see that HCO[1] sets a non-existing feature-gate `WithHostModelCPU` for some reason...
> We can actually implement it.
> 
> I think that we also should fail the migration early ([somewhere here](https://github.com/kubevirt/kubevirt/blob/master/pkg/virt-controller/watch/migration.go#L399)) in the case that no host-model labels were set on a source node.
> [1][https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/controller/operands/kubevirt.go#L108`](https://github.com/kubevirt/hyperconverged-cluster-operator/blob/main/pkg/controller/operands/kubevirt.go#L108%60)

@vladikr Sure, done!
A few questions though:
1. I've changed the name of the variable to `HostModelCPUFeatureGate` to align with our naming convention there. Do you think it's better to call it `WithHostModelCPU` to align with HCO? I tend to think it's more confusing than the first option.
2. Is there an easy way to get the name of this feature gate (`WithHostModelCPU`) from HCO? I'm afraid that if they change the name we won't be aligned.

File: pkg/virt-handler/node-labeller/cpu_plugin.go [hunk 8]
```diff
@@ -77,9 +92,25 @@ func (n *NodeLabeller) loadDomCapabilities() error {
 	}
 
 	usableModels := make([]string, 0)
+	minCpuFeatures := n.getMinCpuFeature()
+	log.Log.Infof("min features: %+v", minCpuFeatures)
```
[the remark this answers is not in the document]
Comment (iholder-redhat) [reply pkg/virt-handler/node-labeller/cpu_plugin.go]: Sure, much better!

File: pkg/virt-controller/watch/migration.go (hunk 7 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:506]: > I think we should fail the migration if we can't find the host model labels on the node. There is nothing really we can wait for if these labels don't exist on the node.

Got you.
Now the migration will fail if host-model labels don't exist for the node (and VMI is using host-model mode), before the migration is scheduled.

Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:506]: > I don't think I understand why these functions are needed?
@iholder-redhat could you please explain?

Sorry, these are leftovers that are now deleted. Ignore them.
(used `nodeInformer` wrong, thanks to @xpivarc it's fixed now)

File: pkg/virt-handler/vm.go [hunk 9]
```diff
@@ -2689,3 +2688,33 @@ func setMissingSRIOVInterfacesNames(interfacesSpecByName map[string]v1.Interface
...
+				vmi.Name, virtconfig.HostModelCPUMigratableFeatureGate)
+		}
+
+		node, err := d.clientset.CoreV1().Nodes().Get(context.Background(), vmi.Status.NodeName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if !nodeHasHostModelLabel(node) {
+			return fmt.Errorf("vmi \"%s\" uses host-model CPU mode and tries to migrate into node \"%s\" "+
```
Comment (vladikr) [inline pkg/virt-handler/vm.go:2705]: I think this error message is a bit confusing. The issue is that we cannot translate what host-model means on this node and therefore we cannot migrate this VMI.
Maybe something like `The node %s has no (%s/...) label to allow migration with host-model.
You could also add a log with a longer message perhaps?

File: tests/migration_test.go [hunk 10]
```diff
@@ -1827,6 +1827,114 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
...
+							return true
+						}
+					}
+				}
+
+				return false
+			}
+
+			It("should migrate only to nodes supporting right cpu model", func() {
+				if !isHeterogeneousCluster() {
```
Comment (vladikr) [inline tests/migration_test.go:1907]: How do we achieve a heterogeneous cluster in kubevirtci?

**Review by vladikr — COMMENTED** [review]
Overall, it looks great! Thank you for making the changes.

I only found one confusing error message and have a question about how did we achieve a heterogeneous cluster in kubevirtci?
/approve

@davidvossel can you please also take a look?

File: tests/migration_test.go (hunk 10 above)
Comment (davidvossel) [reply tests/migration_test.go:1915]: I think we can drop this check and mix up the success criteria a bit so it's accurate for both heterogeneous and homogeneous.

File: pkg/virt-controller/watch/migration.go [hunk 11]
```diff
@@ -1029,3 +1047,41 @@ func (c *MigrationController) findRunningMigrations() ([]*virtv1.VirtualMachineI
...
 	return runningMigrations, nil
 }
+
+func (c *MigrationController) getNodeForVMI(vmi *virtv1.VirtualMachineInstance) (*k8sv1.Node, error) {
+	obj, exists, err := c.nodeInformer.GetStore().GetByKey(vmi.Status.NodeName)
+
+	if err != nil {
+		return nil, fmt.Errorf("cannot get nodes to migrate VMI with host-model CPU. error: %v", err)
+	} else if !exists {
+		return nil, fmt.Errorf("node \"%s\" assosiated with vmi \"%s\" does not exist", vmi.Status.NodeName, vmi.Name)
```
Comment (davidvossel) [inline pkg/virt-controller/watch/migration.go:1057]: typo assosiated/associated

File: pkg/virt-controller/watch/migration.go [hunk 12]
```diff
@@ -1029,3 +1047,41 @@ func (c *MigrationController) findRunningMigrations() ([]*virtv1.VirtualMachineI
...
+
+		if strings.HasPrefix(key, virtv1.HostModelRequiredFeaturesLabel) {
+			requiredFeature := strings.TrimPrefix(key, virtv1.HostModelRequiredFeaturesLabel)
+			pod.Spec.NodeSelector[virtv1.CPUFeatureLabel+requiredFeature] = value
+		}
+	}
+
+	if hostCpuModel == "" {
+		return fmt.Errorf("node does not contain labal \"%s\" with information about host cpu model", virtv1.HostModelCPULabel)
+	}
```
Comment (davidvossel) [inline pkg/virt-controller/watch/migration.go:1082]: I'm hesitant for us to block the migration if hostCpuModel isn't found.

the node labeler isn't always present. for example on arm64 and ppc64 architectures it won't run today. I think it is disabled for emulation mode as well.

does it make sense for this to be a best effort approach instead? If the host cpu model and required features are found, use them for scheduling, and if they aren't then skip?

File: pkg/virt-config/feature-gates.go [hunk 13]
```diff
@@ -28,18 +28,19 @@ const (
...
+	SidecarGate                       = "Sidecar"
+	GPUGate                           = "GPU"
+	HostDevicesGate                   = "HostDevices"
+	SnapshotGate                      = "Snapshot"
+	HotplugVolumesGate                = "HotplugVolumes"
+	HostDiskGate                      = "HostDisk"
+	VirtIOFSGate                      = "ExperimentalVirtiofsSupport"
+	MacvtapGate                       = "Macvtap"
+	DownwardMetricsFeatureGate        = "DownwardMetrics"
+	HostModelCPUMigratableFeatureGate = "HostModelCPUModeMigratable"
```
Comment (davidvossel) [inline pkg/virt-config/feature-gates.go:43]: Does this feature gate disable existing behavior? For example, if someone has a cluster and is migrating VMIs today using the default cpu model, that model is 'host-model' right? So when they update, migrations would no longer work for them until they enable this feature gate which then grants them the previous behavior.

Rather than opting into HostModel migration, since that was the previous default behavior, I'd instead expect a feature gate that allows people to opt-out of host model migration... that's if we need a feature gate at all. I'd be fine with no feature gate.

File: tests/migration_test.go [hunk 14]
```diff
@@ -1835,6 +1835,114 @@ var _ = Describe("[Serial][rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][leve
...
+				// execute a migration, wait for finalized state
+				By("Starting the Migration")
+				migration := tests.NewRandomMigration(vmi.Name, vmi.Namespace)
+				_ = tests.RunMigrationAndExpectCompletion(virtClient, migration, 180)
+
+				newNode, err := virtClient.CoreV1().Nodes().Get(context.Background(), vmi.Status.NodeName, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(isModelSupportedOnNode(newNode, hostModel)).To(BeTrue())
+				expectFeatureToBeSupportedOnNode(newNode, requiredFeatures)
```
Comment (davidvossel) [inline tests/migration_test.go:1943]: In order to guarantee that at least the expected scheduling occurs for both heterogeneous and homogeneous clusters,  you could expand this verification to include cross examining the pod to ensure the right affinity/selectors are set. That would tell us that at least from a scheduling stand point that the host mode migration worked as expected.

**Review by davidvossel — COMMENTED** [review]
I think the PR looks great. The only questions I have are around whether we actually need a feature gate, and if we can slightly modify the functional test's acceptance criteria to work with homogeneous clusters.

File: pkg/virt-controller/watch/migration.go (hunk 12 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:1082]: Hey David!
I'm afraid that if we would do that we will not fix [the bug](https://bugzilla.redhat.com/show_bug.cgi?id=1760028) that this PR aims to fix.

If the VM uses host-model CPU mode and we have no idea which model it is nor which node supports this model then we cannot know which node is valid for scheduling. Today we allow the migration but if the CPU model is not supported on the node the above bug will appear.

As I see it, if we want to always avoid this bug, we have no other option than to fail the migration if we cannot know the actual CPU model.

However, it is reasonable IMO to write in user-guide that in order for VMIs with host-model CPU to migrate correctly node-labeller has to be do its work, and if it doesn't, then we would migrate anyways and hope that we're lucky and the VM lands on a node which supports the right CPU model.

Is that what we want to do? What do you think?

File: pkg/virt-config/feature-gates.go (hunk 13 above)
Comment (iholder-redhat) [reply pkg/virt-config/feature-gates.go:43]: On the one hand I think it makes sense to have this feature gate - in order for VMIs with host-model mode to migrate there's a need for a special care (need to make sure node-labeller runs and that there are nodes in the cluster that do support the same model). Also, HCO uses a similar feature gate for the same reason, as @vladikr [stated above](https://github.com/kubevirt/kubevirt/pull/5623#pullrequestreview-668325105).

On the other hand I understand your point. Backward comparability can be hurt because of this.

The question, IMO, is what's worse:
1. Someone updates Kubevirt and fails to migrate because a new feature gate is turned off by default. In such a case, to fix the problem, the user has to turn on the feature gate.
2. The user updates Kubevirt and everything works as expected. Sometime in the future the system admin adds a new node (or many nodes) that does not support the other nodes' CPU model. The migrations will succeed, but no VM would every migrate to the new node from the old one (and vice-versa). The system administrator might not even notice this for a while and when he does he will probably have to work pretty hard to understand why this is the behavior.

In other words, as I see it, the question is if we want to force our users to raise awareness about this or not. We can force every user that uses such migrations to explicitly declare he understands the requirements, or we can keep things as they are, add documentation, and expect users to read and follow it.

What do you think about this? @vladikr

File: tests/migration_test.go (hunk 14 above)
Comment (iholder-redhat) [reply tests/migration_test.go]: You're absolutely right.
The test is now updated and will occur whether we're running on a homogeneous cluster or not. The host model & node-selector labels will be check for original/target node/pod.

File: pkg/virt-controller/watch/migration.go (hunk 12 above)
Comment (iholder-redhat) [reply pkg/virt-controller/watch/migration.go:1082]: I see your point.
Today, when a VMI with host-model CPU migrates to a node that doesn't support this model the error is the following (as appears in [this bug](https://bugzilla.redhat.com/show_bug.cgi?id=1760028)):
`
{"component":"virt-launcher","kind":"","level":"error","msg":"Live migration failed","name":"vm-fedora","namespace":"default","pos":"manager.go:473","reason":"virError(Code=9, Domain=31, Message='operation failed: guest CPU doesn't match specification: missing features: md-clear')","timestamp":"2019-10-09T15:43:59.828172Z","uid":"2ce9e37d-eaab-11e9-ba41-fa163e35f769"}
`

This means we ask for an invalid operation and depend on libvirt to validate us, which is a bit of a concern. But maybe a warning is enough in case we cannot validate on our own.

P.S. I think it also has something to do with the feature gate discussion on which I've commented below. If we have a feature gate a user that expects migrations of such VMIs is more likely to understand where such an error is coming from.

File: pkg/virt-config/feature-gates.go (hunk 13 above)
Comment (davidvossel) [reply pkg/virt-config/feature-gates.go:43]: for the record, i'd be fine with no feature gate.  The node labeller is part of the kubevirt core, so in the 99% case, everyone will have it working properly.

If anything, as an escape hatch we could add a feature gate that disables the new target node host-model cpu scheduling in the event someone needs to work around something unique in their environment. something like `DisableHostModelCPUScheduling`

Comment (vladikr) [reply pkg/virt-config/feature-gates.go]: The reason why I was in favor of this feature gate is simply that we could quickly reject migrations of VMIs with host-model in the case when the chances for a successful migration are low.
In this case, the user will be able to "correct" the issue by starting the VMI with a different, migratable, CPU model.
This would also prevent multiple migration re-tries during upgrades. 

However, with the current approach, we will not attempt to migrate the VMI if there are no host-model translation labels on the node, so I think we don't need the additional feature gate.

Comment (iholder-redhat) [reply pkg/virt-config/feature-gates.go]: Feature gate is now removed.

**Comment by iholder-redhat** [comment]
They pass locally.
/test pull-kubevirt-unit-test

**Comment by vladikr** [comment]
> They pass locally.
> /test pull-kubevirt-unit-test

Yep, it doesn't pass because the nodeInformer needs more time to add the node object.

File: pkg/virt-controller/watch/migration_test.go [hunk 15]
```diff
@@ -229,6 +236,10 @@ var _ = Describe("Migration watcher", func() {
 		mockQueue.Wait()
 	}
 
+	addNode := func(node *k8sv1.Node) {
+		nodeSource.Add(node)
```
Comment (vladikr) [inline pkg/virt-controller/watch/migration_test.go:240]: As you've seen this is flaky because it needs more time to sync. instead, you can replace  this with something like `nodeInformer.GetIndexer().Add(node)`

**Comment by vladikr** [comment]
Great! Thanks for addressing all the comments. 
/lgtm
